pkg/storage/swarming/swarming_test.go

```diff
@@ -0,0 +1,384 @@
+		fmt.Printf(" -> AlternateSources %d %d %d\n", dev, id, len(as))
+		for i, s := range as {
+			fmt.Printf("     Source%d offset=%d length=%d location=%s hash=%x\n", i, s.Offset, s.Length, s.Location, s.Hash)
+		}
+	}
+
+	prSourceMim.PostSendWriteAtData = func(dev uint32, id uint32, offset int64, data []byte, err error) {
+		fmt.Printf(" -> WriteAtData %d %d offset=%d length=%d %v\n", dev, id, offset, len(data), err)
+	}
+
+	prSourceMim.PostSendWriteAtYouAlreadyHave = func(dev uint32, id uint32, blocksize uint64, blocks []uint32, err error) {
+		fmt.Printf(" -> WriteAtYouAlreadyHave %d %d blocksize=%d %v\n", dev, id, blocksize, err)
+		for _, b := range blocks {
+			fmt.Printf("     Block %d\n", b)
+		}
+	}
+
+	prSourceMim.PostSendWriteAtHash = func(dev uint32, id uint32, offset int64, length int64, hash []byte, loc packets.DataLocation, err error) {
+		fmt.Printf(" -> WriteAtHash %d %d offset=%d length=%d hash=%x loc=%d %v\n", dev, id, offset, length, hash, loc, err)
+	}
+
+	prSourceMim.PostSendEvent = func(dev uint32, id uint32, ev *packets.Event, err error) {
+		fmt.Printf(" -> Event %d %d event=%d/%d data=%d %v\n", dev, id, ev.Type, ev.CustomType, len(ev.CustomPayload), err)
+	}
+
+	prSourceMim.PostRecvEventResponse = func(dev uint32, id uint32, err error) {
+		fmt.Printf(" <- EventResponse %d %d %v\n", dev, id, err)
+	}
+
+	prSourceMim.PostWriteAtResponse = func(dev uint32, id uint32, bytes int, writeErr error, err error) {
+		fmt.Printf(" <- WriteAtResponse %d %d bytes=%d writeErr=%v %v\n", dev, id, bytes, writeErr, err)
+	}
+	/*
+		prSourceMim.PostWaitForPacket = func(dev uint32, id uint32, data []byte, err error) {
+			cmdString := packets.CommandString(data[0])
+			fmt.Printf("  # src-> WaitForPacket dev %d id %d data %d err %v cmd %s\n", dev, id, len(data), err, cmdString)
+		}
+	*/
+	prSourceMim.PostWaitForCommand = func(dev uint32, cmd byte, id uint32, data []byte, err error) {
+		if errors.Is(err, context.Canceled) {
+			return // Don't really care
+		}
+		cmdString := packets.CommandString(cmd)
+		fmt.Printf("  # src-> WaitForCommand dev %d cmd %s id %d data %d err %v\n", dev, cmdString, id, len(data), err)
+	}
+
+	var prDone sync.WaitGroup
+
+	prDone.Add(2)
+	go func() {
+		_ = prSource.Handle()
+		prDone.Done()
+	}()
+	go func() {
+		_ = prDest.Handle()
+		prDone.Done()
+	}()
+
+	// Lets write some data...
+	buff := make([]byte, 4096)
+	for _, s := range testDeviceSchema {
+		prov := dg.GetProviderByName(s.Name)
+		if s.Name == "test1" {
+			allBuff := make([]byte, prov.Size())
+			_, err := rand.Read(allBuff)
+			assert.NoError(t, err)
+			_, err = prov.WriteAt(allBuff, 0)
+			assert.NoError(t, err)
+		} else {
+			for _, offset := range []int64{10000, 300000, 700000} {
+				_, err := rand.Read(buff)
+				assert.NoError(t, err)
+				_, err = prov.WriteAt(buff, offset)
+				assert.NoError(t, err)
+			}
+		}
+	}
+
+	// Make sure it has some time to write to S3
+	time.Sleep(5 * time.Second)
```

Comment:
Same as the comment above.

Reply:
Added TODO 76110c81cafbf9afd291e715fcb317a032c4bce9